bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/ColumnViewer.java

```diff
@@ -817,4 +840,208 @@ protected void setBusy(boolean busy) {
 	public boolean isBusy() {
 		return busy;
 	}
+
+	@Override
+	protected Object[] getSortedChildren(Object parent) {
+		Object[] sorted = super.getSortedChildren(parent);
+		return applyItemsLimit(parent, sorted);
+	}
+
+	/**
+	 * Apply items limit for the items to be created on the viewer. This method is
+	 * supposed to be called at the end of {@link #getSortedChildren(Object)} call,
+	 * which always returns the complete list of elements to be populated.
+	 * <p>
+	 * <ul>
+	 * <li>If the {@link #setItemsLimit(int)} is not used by viewer, this method
+	 * does nothing.</li>
+	 * <li>If the {@link #setItemsLimit(int)} is used by viewer, this method might
+	 * modify the list by reducing number of elements according to the limit set,
+	 * and by adding an {@link IExpandableNode} as last element.</li>
+	 * <li>If the {@link #getSortedChildren(Object)} is not overridden by the
+	 * subclass, this method shouldn't be used by clients.</li>
+	 * </ul>
+	 * Note that in case of parent is {@link IExpandableNode} we will return next
+	 * block of limited elements to be created.
+	 *
+	 * @param parent parent element
+	 * @param sorted all children of given parent, as returned by
+	 *               {@link #getSortedChildren(Object)}
+	 * @return returns only limited items.
+	 * @since 3.31
+	 * @see ColumnViewer#setItemsLimit(int)
+	 */
+	protected final Object[] applyItemsLimit(Object parent, Object[] sorted) {
+		// limit the number of items to be created. sorted always gets the remaining
+		// elements to be created.
+		final int itemsLimit = getItemsLimit();
+		if (itemsLimit <= 0) {
+			return sorted;
+		}
+
+		if (sorted.length <= itemsLimit) {
+			return sorted;
+		}
+
+		int offSet = itemsLimit;
+		int srcPos = 0;
+
+		if (parent instanceof IExpandableNode expNode) {
+			srcPos = expNode.getOffset();
+			offSet = srcPos + itemsLimit;
+			// pass on original children
+			sorted = expNode.getAllElements();
+		}
+
+		Object[] partialChildren = new Object[itemsLimit + 1];
+
+		// Extract a subset of children
+		System.arraycopy(sorted, srcPos, partialChildren, 0, itemsLimit);
+
+		// Add an expandable node
+		partialChildren[itemsLimit] = createExpandableNode(sorted, offSet, itemsLimit);
+
+		return partialChildren;
+	}
+
+	/**
+	 * Concrete viewer is supposed to "expand" {@link IExpandableNode} located at
+	 * given cell.
+	 * <p>
+	 * Default implementation does nothing.
+	 *
+	 * @param cell selected on click
+	 * @since 3.31
+	 */
+	protected void handleExpandableNodeClicked(Widget cell) {
```

Comment:
Would it be possible to keep is package protected (thus not API)?

Reply:
fixed.